src/arviz_stats/metrics.py

```diff
@@ -105,3 +107,270 @@ def r2_score(
+           ...: data2 = load_arviz_data('non_centered_eight')
+           ...: wasserstein(data1, data2, var_names="mu")
+
+    References
+    ----------
+
+    .. [1] "Wasserstein metric",
+           https://en.wikipedia.org/wiki/Wasserstein_metric
+    """
+    dist1, dist2 = _prepare_distribution_pair(
+        data1,
+        data2,
+        group=group,
+        var_names=var_names,
+        sample_dims=sample_dims,
+        num_samples=num_samples,
+        random_seed=random_seed,
+    )
+
+    if joint:
+        distance = wasserstein_distance_nd(dist1, dist2)
+
+    else:
+        distance = 0
+        for var1, var2 in zip(dist1.T, dist2.T):
+            distance += wasserstein_distance(var1, var2)
+
+    return distance
+
+
+def _prepare_distribution_pair(
+    data1, data2, group, var_names, sample_dims, num_samples, random_seed
+):
+    """Prepare the distribution pair for metric calculations."""
+    data1 = convert_to_datatree(data1)
+    data2 = convert_to_datatree(data2)
+    if sample_dims is None:
+        sample_dims = rcParams["data.sample_dims"]
+
+    dist1 = _extract_and_reindex(
+        data1,
+        group=group,
+        var_names=var_names,
+        sample_dims=sample_dims,
+        num_samples=num_samples,
+        random_seed=random_seed,
+    )
+
+    dist2 = _extract_and_reindex(
+        data2,
+        group=group,
+        var_names=var_names,
+        sample_dims=sample_dims,
+        num_samples=num_samples,
+        random_seed=random_seed,
+    )
+
+    shared_var_names = set(dist1.data_vars) & set(dist2.data_vars)
+    if not shared_var_names:
+        raise ValueError(
+            "No shared variable names found between the two datasets. "
+            "Ensure that both datasets contain variables with matching names."
+        )
+
+    if var_names is None:
+        var_names = list(shared_var_names)
+        dist1, dist2 = dist1[var_names], dist2[var_names]
+
+    dist1 = dataset_to_dataframe(dist1, sample_dims=["sample"]).values
+    dist2 = dataset_to_dataframe(dist2, sample_dims=["sample"]).values
+
+    return dist1, dist2
+
+
+def _extract_and_reindex(data, group, var_names, sample_dims, num_samples, random_seed):
+    return (
+        extract(
+            data,
+            group=group,
+            var_names=var_names,
+            num_samples=num_samples,
+            random_seed=random_seed,
+            keep_dataset=True,
+        )
+        .reset_index("sample")
+        .drop_vars(sample_dims)
+        .assign_coords(sample=range(num_samples))
+    )
+
+
+def _kld(ary0, ary1):
+    """Kullback-Leibler divergence approximation.
+
+    Compute KL-divergence using equation 14 from [1]_. Assumes both arrays
+    are of the same shape.
+
+    Parameters
+    ----------
+    ary0 : array-like
+        First sample of shape (n_posterior_samples, n_outputs).
+    ary1 : array-like
+        Second sample of shape (n_posterior_samples, n_outputs).
```

Comment:
```suggestion
    ary0, ary1 : (N, M) array-like
        Samples of the input distributions. ``N`` represents the number of samples (e.g. posterior samples)
        and ``M`` the number of outputs (e.g. number of variables in the posterior)
```

It is common to include the shape information in the type, and being same type and shape we can also share type and description between the arguments